### ISSUE TITLE
conditional SDs and se.fit of predictions

### ISSUE THREAD

**Issue by bbolker** [lme4/lme4#739]
Emi Tanaka and I have been doing some work (with Francis Hui) on conditional SDs and covariances between random and fixed-effect parameters. Although @dmbates has long contended that covariances between random and fixed effects don't mean anything (since one is a random variable and one is a parameter), it is indeed possible to recover these values, and hence to compute standard errors of predictions that condition only on $\hat \theta$ (the estimated covariance parameters), incorporating uncertainty both in the conditional modes and in the fixed-effect estimates. We have a working implementation of this in the `predict` method; the user can now request `se.fit = TRUE` for predictions.

There is more work to be done (e.g. comparisons with `glmmTMB` results, which by default give uncertainty that doesn't condition on $\hat \theta$ either, but could be hacked to give SEs conditional on $\hat \theta$; simulations + coverage experiments), but I wanted to bring it up as a placeholder/before we got too much farther.  (We have some documentation but it's currently sitting on a private GH repo elsewhere ...)

**Comment by mmaechler** [comment]
Well, there must be an intelligible probability/msth-statistics definition of what the "values" here are.
I assume  they are *predictions* (of a random variable) of some conditional expectations  E[... | ...], or ???
If that's not possible to specified, I'd stay with Doug @dmbates  that I'd hate us computing values of something that is only defined by a numerical algorithm ...

**Comment by bbolker** [comment]
This is a passage from Welham, Sue, Brian Cullis, Beverley Gogel, Arthur Gilmour, and Robin Thompson. 2004. “Prediction in Linear Mixed Models.” Australian & New Zealand Journal of Statistics 46 (3): 325–47. https://doi.org/10.1111/j.1467-842X.2004.00334.x.

I don't know if they're just papering over theoretical difficulties, but they seem perfectly happy working with the joint covariance matrix of the fixed and random effects ...

![Screenshot from 2023-09-04 11-20-19](https://github.com/lme4/lme4/assets/78918/077e8fe4-10a7-4df8-bd81-92068623800b)

**Comment by bbolker** [comment]
I'm hoping that Francis Hui might chime in on this (he's in Australia, so not until tonight/tomorrow anyway), as he has a PhD student working on related issues ...

**Comment by bbolker** [comment]
> I'm actually not sure why they assert that this is the measure to use instead of $var(\tilde{\boldsymbol{\beta}})$

¿¿ maybe they're considering the possibility of bias in $\tilde{\boldsymbol{\beta}}$, i.e. they want the equivalent of MSE rather than variance? Or maybe the expected value of $\tilde\boldsymbol{\beta}}$ is $\boldsymbol{\beta}}$ and these are actually equivalent?? (I'm not sure, I haven't thought this through too carefully.)

**Comment by emitanaka** [comment]
Just a note that $var(\tilde{\boldsymbol{\beta}}|\boldsymbol{\theta})$ won't be equal to $var(\tilde{\boldsymbol{\beta}} - \boldsymbol{\beta}|\boldsymbol{\theta})$ (I recall doing the maths somewhere to check this).

**Comment by emitanaka** [comment]
Or to be more specific, it's only equal iff $cov(\tilde{\boldsymbol{u}}, \boldsymbol{u}) = 0$ (which is not guaranteed).

**Comment by fhui28** [comment]
Apologies for the slowness on my end folks!

My student Nickson @u5838836 can comment more and correct me, but here is my take on the discussion so far.

Let us assume the (co)variance components $\theta$ are known for now, to keep things manageable [or equivalently, 
condition on $\hat{\theta}$ as per @bbolker mentioned]. Then to be honest, I am not sure I understand the conceptual concern here: both the TMB formula and the "Bolker-Tanaka" method (which is based on inverting the Hessian of the joint log-likelihood of the GLMM) aim to estimate the same thing, namely try to estimate $Cov((\hat{\beta}, \hat{b}))$ i.e., the covariance matrix of the vector of estimated fixed and random effects $(\hat{\beta}, \hat{b})$. Then `se.fits` of predictions are extracted from big covariance matrix.

To me, this quantity is intelligible and mathematically sound: everything in $(\hat{\beta}, \hat{b})$ is an estimate [well one is strictly a predictor] and so has stochasticity through being functions of $y$. So calculating the covariance matrix, and thus calculating the cross-covariance between $\hat{\beta}$ and $\hat{b}$ looks fine. 

If we had issues with wanting to estimate this covariance matrix, then we also would have to call into question whether what TMB does makes sense; see also [this paper which is an attempt to derive the TMB formula more formally](https://doi.org/10.1016/j.csda.2021.107227).

**Comment by fhui28** [comment]
Two extra tidbits.

1. While the TMB formula and the Bolker-Tanaka method both aim to estimate the same thing, and indeed end up with the same formula as far as I can tell (at least in the case assuming $\theta$ is known), the way they can here is quite different in terms of derivations and assumptions. 
2. Putting the concepts and maths aside, in practice does estimating $Cov((\hat{\beta}, \hat{b}))$ and then basing `se.fits` of predictions from this work? That is, can you then slap on a normal distribution and use those as prediction intervals? The answer does seem to be yes, although the reason why is more complex than it first seems. I won't delve into it now, and Nickson @u5838836  can dive if he wanted to. 
3. And just to be clear, the above does works much better compared to what the CRAN version of `lme4` currently does, which I think is justified along the grounds of wanting to approximate the covariance of f(b|y) and all parameters assumed to be known.

**Comment by u5838836** [comment]
Hi everyone, some of my thoughts below (let's assume $\theta$ is known):

1) I believe TMB [tries to] get the (unconditional) covariance matrix of $\tilde{\beta} - \beta$. As Francis says, there is no conceptual problem as $\tilde{\tau}$ is a random variable and thus so is $\tilde{\tau} - \tau$; $\tilde{u}$ is a random variable and so is $u$ (and they are not perfectly correlated), and thus so is $\tilde{u} - u$. The covariance of interest in this discussion would then have the interpretation of the covariance between these two random variables $\tilde{\tau} - \tau$ and $\tilde{u} - u$ over repeated realisations of the data generating process/model. As Emi mentioned, the covariance matrix of $\tilde{\beta} - \beta$ is in general not the same as that of $\tilde{\beta}$ because $u$ is a random variable.
2) It's important the measure is for the gap/difference $\tilde{\beta} - \beta$ rather than just the estimators $\tilde{\beta}$, in particular if we want to use the measure to build prediction intervals for functions of the true/realised random effects. For example, if we knew the distribution of $\hat{u} - u$ we could use the quantiles of said distribution and pivot to obtain prediction bounds for $u$. However, if we knew only the distribution of $hat{u}$ we would be unable to build a prediction interval for $u$ because $u$ does not appear anywhere in the distribution of $\hat{u}$ (which can only depend on parameters, not random variables such as $u$).
3) The covariance between $\tilde{\tau} - \tau$ and $\tilde{u} - u$ is important.  If this covariance is ignored, the correct variance for $(\hat{\tau} + \hat{u}_i) - (\tau - u_i)$ [and therefore predictions for $y$ or $\mu$, which itself is a random variable] for example will not be obtained. The way glmmTMB does things currently does incorporate this covariance.
4) Prediction intervals for $y$ is a different beast for GLMMs and distinct from prediction intervals for $\mu$; for LMMs maybe one just needs to add on the variance for epsilon on top of the variance for $\hat{\mu} - \mu$ and that's enough.

**Comment by dmbates** [comment]
Much of this discussion is predicated on considering the conditional modes of the random effects as parameter estimates but,  as @fhui28 writes, "... everything in $(\hat\beta, \hat{b})$ is an estimate [well one is strictly a predictor] ...".

That bracketed comment is an important distinction to me.  I was a mathematician before I started a Ph.D. program in statistics and I still think like a mathematician.  In this case that means that I want to at least have the possibility of deriving the quantities of interest from the probability model.  There is a lot of literature that purports to do so but bear in mind that not everything written in the literature stands up to the test of time.  Much of what was written about random effects and BLUPs and BLUEs was rather casual about definitions and derivations and doesn't stand up to close scrutiny.  (We're even back to discussion of ideas like negative variances on the R-SIG-Mixed-Models mailing list.)

Let's consider the linear mixed-effects model for which the probability model can be written with the unconditional distribution of the random effects vector, $\mathcal{B}\sim\mathcal{N}\left(\mathbf{0}, \boldsymbol{\Sigma}\right)$, and the conditional distribution of the response vector $\mathcal{Y}$, given $\mathcal{B}=\mathbf{b}$, which is $\mathcal{N}(\mathbf{X}\boldsymbol{\beta}+\mathbf{Zb},\sigma^2\mathbf{I})$.

In practice both $\mathbf{Z}$ and $\boldsymbol{\Sigma}$ are large but sparse and highly structured.  That doesn't affect the theory but it is important in creating feasible computational methods.  At this point we will just say that $\boldsymbol{\Sigma}$ is positive semi-definite and is determined by a small vector of parameters $\boldsymbol{\theta}$.  (We modify that for computational convenience but that doesn't affect the theoretical result.)

Given these two distributions we form the joint density of $(\mathcal{Y}, \mathcal{B})$ and the marginal distribution of $\mathcal{Y}$ (by integrating the joint density w.r.t. $\mathcal{B}$).  These are probability distributions in which the parameters, $\boldsymbol{\beta}$, $\sigma^2$, and $\boldsymbol{\theta}$ are considered to be known and fixed.  The likelihood for a particular set of values of the parameters is the marginal density of $\mathcal{Y}$ evaluated at the observed responses,  $\mathbf{y}$, and those parameters.  The maximum likelihood *estimates* of the parameters are the values of the parameters that maximize the (log-)likelihood.  The maximum likelihood *estimator* is the random variable defined by the definition of the m.l.e.'s  w.r.t. the marginal distribution of $\mathcal{Y}$, usually evaluated at the estimated parameter values.  In the definition of the estimator the effect of the random variable $\mathcal{B}$ has been integrated out of the marginal distribution of $\mathcal{Y}$.

I don't want to come off as too pedantic but I think it is important to distinguish between estimates and the estimator here.  What we may describe as covariance of estimates is actually correlation of estimators.  In this case I don't see how I would create a covariance of the estimator of $\boldsymbol{\beta}$ with the random variable $\mathcal{B}$, because $\mathcal{B}$ has been integrated out of the definition of the estimator.

As I understand the distribution of the predictor of $\mathcal{B}$, it is the conditional distribution of $\mathcal{B}$ given $\mathcal{Y}=\mathbf{y}$ evaluated at the parameter estimates.  Again, I don't see how I could use this definition to create a covariance of $\mathcal{B}$ and the parameter $\boldsymbol{\beta}$, which is fixed.

This is why I can't make sense of the covariance of the estimator of $\mathbf{\beta}$ and the predictor of $\mathcal{B}$. I can't formulate a joint distribution for these two quantities.

OF course, things become very easy in a Bayesian formulation because parameters are random variables so it is straightforward to have a joint distribution of $\boldsymbol{\beta}$ and $\mathcal{B}$.

**Comment by u5838836** [comment]
I think the difference lies in the way we view predictors. 

-  I agree there is no relationship between the estimator of $\boldsymbol{\beta}$ and $\mathcal{B}$ when we condition on $\mathcal{Y}=\mathbf{y}$, because the estimator of $\boldsymbol{\beta}$ is then just a constant and there is no covariance to be had between the estimator of $\boldsymbol{\beta}$ and $\mathcal{B}$.
- However, that is fine and doesn't contradict what Francis & I said previously - because we are not interested in a conditional on $\mathcal{Y}=\mathbf{y}$ sampling framework, because we are not Bayesian.

The way we currently understand/think about things that allows us to interpret a covariance between $\boldsymbol{\beta}$ and $\mathcal{B}$ is as follows:

- As soon as we pick a predictor of $\mathcal{B}$ e.g., the BLUP, then from a frequentist point of view the conditional distribution $\mathcal{B} |  \mathcal{Y}=\mathbf{y}]$ is no longer relevant. The only thing we can talk about, and the only that matters, is the sampling distribution of the BLUP, and from this perspective it makes sense to talk about the covariance between $\boldsymbol{\beta}$ and this predictor. 
- Moreover, the sampling distribution of the BLUP in no way has to relate to $\mathcal{B} |  \mathcal{Y}=\mathbf{y}]$. In fact, arguably and perhaps controversially, the two can not be related: in $[\mathcal{B} |  \mathcal{Y}=\mathbf{y}]]$ how can you talk about repeated sampling of the responses if you condition on all the responses. 
- It is natural to want to focus on the distribution of $\mathcal{B} |  \mathcal{Y}=\mathbf{y}]$ because in LMMs we derive BLUPs using it, in a separate procedure to the (MLE ) estimation of the parameters. But actually, it's just a means to an end; for example in PQL or Laplace estimation of GLMMs, we simultaneously get the estimators and predictors, without strictly needing to involve the distribution of $\mathcal{B} |  \mathcal{Y}=\mathbf{y}]$ anywhere. There are also other ways the BLUP can be derived.
- In other words: we see estimators and predictors both as just functions of the data $\mathcal{Y}$, so they will have a covariance when working in an unconditional framework (where $\mathcal{Y}$ is a random variable). 
- That is, in this case, we see $E[\mathcal{B} |  \mathcal{Y}]$, or the mode of $f(\mathcal{B} |  \mathcal{Y})$ for example, as 1) point predictors of \mathcal{B}, 2) functions of $\mathcal{Y}$, and 3) random variables when unconditional on \mathcal{Y} (which is the sampling framework we always work in as frequentists). They are not distributions.

**Comment by u5838836** [comment]
Apologies for not being clear. An estimator is a random variable derived from the sample random variable. But if we consider a sampling framework where we condition on a particular realisation of $\mathcal{Y}$, say $\mathbf{y}$, then $\mathcal{Y}$ is no longer a random variable and neither is any estimator. Meanwhile however, the random effects $\mathcal{B}$ still has a non-degenerate distribution $f(\mathcal{B} | \mathbf{y})$ and is a random variable even when our sampling framework conditions on $\mathcal{Y} = \mathbf{y}$. I don't think this point is too important or directly relevant though, because as frequentists we don't care about a sampling framework that conditions on $\mathcal{Y} = \mathbf{y}$. One consequence of this is that quantities such as $var(\mathcal{B} | \mathcal{Y} = \mathbf{y})$ will not be of direct interest.

To summarise my main points though, I guess it would be

1. An estimator of the fixed effects is some function of the random variable $\mathcal{Y}$. A predictor of the random effects is some (probably different) function of the random variable $\mathcal{Y}$. Thus they (the estimator and predictor) have a joint distribution and covariance as long as the marginal distribution of $\mathcal{Y}$ is not degenerate.
2. An estimator can have a joint distribution with the random effects $\mathcal{B}$ even if the random effects are integrated out in the likelihood, in the same way $\mathcal{Y}$ has a joint distribution with $\mathcal{B}$

- Let's take a stylised example that I think illustrates some of my points. Consider the model $\mathcal{Y}_{ij}$ = \beta + \alpha_i + \epsilon_{ij}$ for $i = 1,...,n$ and $j=1,...,m$, where $\alpha_i$ and $\epsilon_{ij}$ are all i.i.d. standard normal. 
- Let's define a (stupid) estimator of $\beta$, $\hat{\beta} = \mathcal{Y}_{11}$. Then $\hat{\beta}$ and $\alpha_1$ are multivariate normal with a covariance of $var(\alpha_1) = 1$, and are independent of $\alpha_i$ for $i=2,...,n$. 
- Now let's define a stupid predictor of $\alpha_1$ as $\hat{\alpha}_1 = 0.5 \mathcal{Y}_{12}$. Then $\hat{\beta}$ and $\hat{\alpha}_1$ are multivariate normal with a covariance of $cov(\mathcal{Y}_{11}, 0.5\mathcal{Y}_{12}) = 0.5 var(\alpha_1) = 0.5 $. 
- $\hat{\beta} - \beta$ and $\hat{\alpha}_1 - \alpha_1$ are multivariate normal with covariance $cov( \mathcal{Y}_{11} ,  \mathcal{Y}_{12} - \alpha_1) = -0.5$.

**Comment by dmbates** [comment]
I agree with what @u5838836 has written about both of these quantities being functions of $\mathcal{Y}$, which also gives us a convenient way to evaluate these covariances.  Equations (17) and (18) in the JSS paper that Ben and Martin and Steve and I wrote (https://doi.org/10.18637/jss.v067.i01) describe solving for the mean of $\mathcal{U}$ where $\mathcal{B}=\boldsymbol{\Lambda}\mathcal{U}$ and $\hat{\boldsymbol{\beta}}$ simultaneously.  Thus the covariance of the conditional mean of $\mathcal{U}$ (and, hence $\mathcal{B}$) and the estimator of $\boldsymbol{\beta}$ can be straightforwardly derived from $\mathbf{L}$, $\mathbf{R}_{ZX}$, and $\mathbf{R}_X$.  It will mean forming the inverse of a large, sparse, triangular matrix which could end up being a large, dense, triangular matrix but that is the nature of the calculation I think.

**Comment by bbolker** [comment]
@dmbates, Emi already implemented a reasonably sophisticated version of what you're describing [here](https://github.com/lme4/lme4/blob/master/R/predict.R#L532-L546), which uses the fact that the matrix formed of $\mathbf L$, $\mathbf R_{ZX}$, $\mathbf R_X$ is block-lower triangular, so we can form the inverse from (functions of) the inverses of the blocks &mdash; and we already have the Cholesky factors of the blocks, so getting the inverses is relatively straightforward.  It's true that the inverse is dense, and could be big (if we e.g. have but we don't have to do very much with it ...

A quick benchmark suggests this will be OK up to a few 1000 clusters, but maybe problematic beyond that ...

```r
library(lme4)
library(peakRAM)
bfun <- function(npergrp=5, ngrp = 2000) {
    set.seed(101)
    bigdat <- data.frame(x = rnorm(npergrp*ngrp),
                         f = factor(rep(1:ngrp, each = npergrp)))
    bigdat$y <- simulate(~x + (x|f),
                         newparams = list(beta = c(1,1),
                                          theta = c(2,0,1),
                                          sigma = 1),
                         family = gaussian,
                         newdata = bigdat)[[1]]
    bigfit <- lmer(y~x + (x|f),
                   data = bigdat)
    predict(bigfit, se.fit = TRUE)
}
peakRAM(bfun(ngrp = 1000))
peakRAM(bfun(ngrp = 2000))
peakRAM(bfun(ngrp = 3000))
```

```r
  Function_Call Elapsed_Time_sec Total_RAM_Used_MiB Peak_RAM_Used_MiB
1 bfun(ngrp=1000)            0.646                0.1             510.3
1 bfun(ngrp=2000)            3.208                0.2            1625.4
1 bfun(ngrp=3000)            7.114                0.4            3651.2
```

**Comment by WillTheKiwi** [comment]
Someone alerted me to this conversation, so I have just signed up to ask what might be a stupid question: does this conversation relate to the covariance between the intercept (random effect) and the slope (fixed effect, sort of, except it's not, really) in a random intercepts and slopes model? If yes, there is no problem in interpretation of the covariance, when you divide it by the intercept variance: it's the modifying effect of the intercept on the slope. If not, can someone provide a concrete example of how you can have a covariance between two variables, one of which is fixed and one of which is random? They would both have to have the same number of levels and somehow represent different things, because if they represent the same thing, they both have to be either fixed or random.

**Comment by bbolker** [comment]
@WillTheKiwi, the conversation is about the covariances between fixed-effect parameters ($\beta$ values in the standard notation) and random-effect values ("BLUPs" or "conditional modes") ($b$ values in the notation of the JSS paper/`lmer` vignette, sometimes $u$ in other notations). We are interested in this question because we want to be able to estimate the (conditional?) variance of predicted quantities: for example, in a simple model `y ~ 1 + (1|g)` (intercept plus random intercepts), the predicted mean for observations from group $i$ is $\beta_0 + b_i$. If we don't know (or can't define) the covariance between $\beta_0$ and $b_i$, it's hard to find a reliable value for the variance.

**Comment by emitanaka** [comment]
@WillTheKiwi also a small note that the above discussion isn't about $cov(\boldsymbol{\beta}, \boldsymbol{b})$ (the covariance of fixed and random effects) but $cov(\hat{\boldsymbol{\beta}}, \tilde{\boldsymbol{b}})$ the covariance of BLUEs and BLUPs.

**Comment by fhui28** [comment]
Something Nickson @u5838836 and I have not quite worked out yet (I think) is why the proposed method for calculating standard errors of prediction actually works. That is, we seem to be in some agreement now with wanting to use the Hessian of the joint log-likelihood of the GLMM as the new approach to deploy in `lme4` for `se.fit`, but what's the intuition? 

> while the TMB formula and the Bolker-Tanaka method both aim to estimate the same thing, and indeed end up with the same formula as far as I can tell (at least in the case assuming is known), the way they can here is quite different in terms of derivations and assumptions.

The TMB formula at least has some theoretical motivation/development behind it. Using the Hessian of the joint log-likelihood of the GLMM currently does not as far as I can tell (and it would be nice to perhaps hint at why this works in the package documentation)? That being said, I hope to have a chat with @emitanaka later this week to understand why this approach works for LMMs, and from there we might be able to generate a motivation/"derivation" for GLMMs.

**Comment by bbolker** [comment]
@dmbates and @mmaechler , can I assume that you are now comfortable enough with the theoretical framework being proposed (as much as we've worked out) that it's OK to proceed farther with integrating this approach into `lme4` (given I/we can find the time and energy ...) ?

**Comment by fhui28** [comment]
I had a chat with @emitanaka and reckon we have managed to come up with yet another derivation for the Bolker-Tanaka method of calculating prediction standard errors (which may be easier to incorporate into a help file explanation or something in `lme4`). The approach is basically to get it by _brute force_: 

In LMMs $y = X\beta + Zu + e$, $\hat{\beta} = (X^\top V^{-1} X)^{-1} X^\top V^{-1} y$ with $V$ being the marginal covariance, while the predicted random effect is $\hat{u} = G Z^\top V^{-1}(y - X\hat{\beta})$ where $G$ is the random effects covariance matrix. Therefore, assuming the variance components are known, you can then directly calculate the target of interest i.e., the covariance matrix of $Cov((\hat{\beta} - \beta, \hat{u} - u))$. Standard errors of predicted values can then be constructed from this. Note this is more or less the same as what @bbolker posted a while back with the paper <https://doi.org/10.1111/j.1467-842X.2004.00334.x>

For GLMMs, what we can do is to say that at convergence the GLMM solves the approximate LMM $\tilde{y} = X\beta + Zu + e$ where $\tilde{y}$ is a set of working responses similar to what happens in GLMs, and $e$ is a vector of approximate error terms with mean zero and covariance given a diagonal matrix of weights (again similar to what happens in GLMs). From this approximate LMM, you can then follow the same process as for the LMM above i.e., get the estimator $\hat{\beta}$ and $\hat{u}$ and then apply brute force algebra to compute $Cov((\hat{\beta} - \beta, \hat{u} - u))$. Assuming variance components are known, this will lead to the same form as the TMB formula, or equivalently the inverse negative Hessian of the joint log-likelihood of the GLMM.

Of course like @u5838836's derivation, there are hand-wavy parts here with the above brute force approach, namely that you treat the GLMM as an approximate LMM at convergence (note methods like PQL basically do this anyway to fit the GLMM), and related that the estimators for $\beta$ and $u$ from approximate LMM are the same as those based on using `lme4` or `glmmTMB`. But I personally feel like this is easier to explain as a motivation for why this new method for calculating prediction standard errors.

**Comment by mmaechler** [comment]
>  assuming the variance components are known

is crucial ... and I think points to the big elephant here:  If the $\theta_j$s are not known as in reality, and hence estimated, they contribute an extra level of noise.   Consequently, you know that all your intervals (determined by the above assumption)  are too small -- maybe only a little bit, maybe largely -- but you don't easily see by how much they are too small.

**Comment by bbolker** [comment]
@mmaechler, fair enough: but we frequently use "good enough" approximations. (For example, `summary.glm()` reports p-values without finite-size corrections ...) I still think that what's being discussed here is an improvement on the current implementation (i.e., conditioning on y and theta is still better than conditioning on b and theta ...) 

We could do some tests/examples comparing coverage of various ways of computing CIs: parametric bootstrapping, glmmTMB analogue (uses joint uncertainty *including* theta), conditioning on b+theta, conditioning on y+theta ... with simulated examples we can assess coverage ... (this is of course some amount of work, and will be restricted in scope to the examples we choose ...)

**Comment by fhui28** [comment]
Fair enough points raised by @mmaechler and @bbolker. Of course assuming $\theta$ is known when it has to be estimated in practice should be [?!] not an issue asymptotically. So we are talking finite sample size elephant which can be quite big if the sample size is small. 

On the "good enough" approximations, so assuming $\theta$ is done even in point prediction e.g., to go from BLUPs to E-BLUPs in LMMs are replace the unknown $\theta$ with $\hat{\theta}$. "Putting the E on" actually induces a finite sample bias.
